rfcs/0012-external-function.md

```diff
@@ -0,0 +1,163 @@
+```
+
+However, expressions with short-circuit optimization should evaluate their arguments serially.
+
+```sql
+SELECT udf1(x) && udf2(y) && udf3(z) FROM t;
+```
+
+#### pg_sleep?
+
+BTW, previously, we removed a built-in async expression `pg_sleep`. We won't add it back even after implementing the async expression framework because its semantics will break many concurrency possibilities.
+
+### Cross-chunk concurrency
+
+Concurrency between expressions is not enough for some cases, so we should also handle chunks concurrently.
+
+Thanks to the Stream-based execution model, we can quickly achieve the result by [`Stream::buffered`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffered).
+
+### Error handling
+
+We have no choice but to retry. One way to minimize the loss is to introduce the partial-checkpoint mechanism.
+
+## Multiple return values
+
+PostgreSQL allows users to specify multiple return values:
+
+```sql
+CREATE FUNCTION sum_n_product (x int, y int, OUT sum int, OUT product int)
+AS 'SELECT x + y, x * y'
+LANGUAGE SQL;
+```
+
+We can refer to the syntax to support multiple return values but pack them in an anonymous struct.
+
+### Deterministic
+
+We expect all UDFs will output a deterministic result. However,  some functions are not deterministic. There are three accepted ways:
+
+Only allow UDFs over an append-only stream.
+Force a checked `MaterializeExecutor` over the UDF.
+Force UDFs to be explicitly defined with the DETERMINISTIC keyword and trust the users.
+
+BTW we can also apply the solution to built-in non-deterministic functions, such as `random(),` `proc_time()` or `now()`.
+
+## Alternatives
+
+### Introduce a new operator, `ProjectAsync` to execute async UDFs
+
+The solution heavily invades the planner part.
+
+We can plan the SQL as the following plan:
+
+```sql
+SELECT round(my_udf(to_char(x))) FROM t;
+```
+
+```plain
+LogicalProject { exprs: [round($0)] }
+└─LogicalAsyncProject { exprs: [my_udf($0)] }
+  └─LogicalProject { exprs: [to_char($0)] }
+    └─LogicalScan { table: t, columns: [t.x] }
+```
+
+The solution can work well on `Project`, `Filter`, and `Join`, but failed on [the filter clause](https://medium.com/little-programming-joys/the-filter-clause-in-postgres-9-4-3dd327d3c852). There are also many other corner cases since expressions are used everywhere in PostgreSQL.
+
+Also, this is not an elegant solution. For the frontend part, `Project` and `ProjectAsync` are the same.
+
+## Future possibilities
+
+### Using another RPC framework
+
+The external function framework will not be public to users, so we use gRPC in our first version for simplicity. We can investigate other RPC frameworks later.
+
+### Using wasm-based implementation and embedding the VM into the kernel?
+
+A good idea, but currently, the capability of the wasm is limited. For some CPU-bounded functions written in some languages without heavy runtime, it's a better solution for performance.
```

Comment:
> the capability of the wasm is limited

Can you give some examples? 🤔

Reply:
For me there are two points:
1. Currently there is no stable solution to convert java code to wasm
2. It seems that wasi is not fully supported.